contrib/babelfishpg_tsql/src/tsql_for/forjson.c

```diff
@@ -571,3 +897,147 @@ insert_existing_json(JsonbValue *current, JsonbValue* parent, JsonbValue *nested
     // update parent pointer
     parent->val.object.pairs[idx].value = *current;
 }
+
+// This function adds a json pair to a given array
+static void
+insert_existing_json_to_obj(JsonbValue *current, JsonbValue* parent, JsonbValue *nestedVal, int idx, char *key)
+{
+	JsonbValue *jsonbArray;
+	JsonbValue *jsonKey;
+	JsonbPair* newPair;
+	JsonbPair* currPairs;
+	JsonbPair* newPairs;
+	int numPairs;
+
+	jsonKey = palloc(sizeof(JsonbValue));
+	jsonKey->type = jbvString;
+	jsonKey->val.string.len = strlen(key);
+	jsonKey->val.string.val = pstrdup(key);
+
+	newPair = palloc(sizeof(JsonbPair));
+	newPair->key = *jsonKey;
+	newPair->value = *nestedVal;
+
+	jsonbArray = (JsonbValue *) (((JsonbPair*) current->val.object.pairs)->value.val.array.elems); // Object within the array
+	currPairs = (JsonbPair*) jsonbArray->val.object.pairs;
+	numPairs = jsonbArray->val.object.nPairs;
+
+	// Allocate space for the new pairs
+	newPairs = palloc(sizeof(JsonbPair) * (numPairs + 1));
+
+	for(int i = 0; i < numPairs; i++)
+		newPairs[i] = currPairs[i];
+	newPairs[numPairs] = *newPair;
+
+	jsonbArray->val.object.nPairs = numPairs + 1;
+
+	jsonbArray->val.object.pairs = newPairs;	
+
+	parent->val.object.pairs[idx].value = ((JsonbPair*) current->val.object.pairs)->value;
+
+	return;
+}
+
+/*
+ * checkForDuplicateRows inserts the given row into the json array 
+ * nested based on the root object
+ */
+static void
+checkForDuplicateRows(JsonbValue *jsonbArray, JsonbValue* row, int maxDepth, int currDepth, int* minInsertDepth)
+{
+	JsonbPair *arrRowPairs;
+	JsonbPair *rowPairs;
+	if(currDepth == maxDepth) {
+		jsonbArray->val.array.nElems++;
+		jsonbArray->val.array.elems = (JsonbValue *) repalloc(jsonbArray->val.array.elems, sizeof(JsonbValue) * (jsonbArray->val.array.nElems));
```

Comment:
Why we realloc some space in here ?

Reply:
We need to allocate space for the new array element being added to jsonbArray